File: week-1/problem-set-recursion/prob2.py
"""Prob2 example.

This file belongs to the week 1 recursion practice solutions. It uses find_max to illustrate the main idea with a small runnable example.
"""

def find_max(numbers):
    """Return the largest value in the list using recursion.

    Args:
        numbers (list): The list of numbers used in the example.

    Returns:
        int: The integer result produced by the algorithm.
    """
    if len(numbers) == 1:
        return numbers[0]
    else:
        max_val = find_max(numbers[1:])
        return numbers[0] if numbers[0] > max_val else max_val

numbers = [5, 10, 3, 8, 1]
print(find_max(numbers))

File: week-1/problem-set-recursion/prob2.cpp
/**
 * @file prob2.cpp
 * @brief C++ version of the recursive maximum-search example.
 */

#include <algorithm>
#include <iostream>
#include <vector>

using namespace std;

/**
 * @brief Return the largest value in a list using recursion.
 * @param numbers const vector<int>& Input list of integers.
 * @param index size_t Current position examined by the recursive call.
 * @return int Largest value found from index to the end of the list.
 */
int find_max(const vector<int>& numbers, size_t index = 0) {
    if (index + 1 == numbers.size()) {
        return numbers[index];
    }
    int max_val = find_max(numbers, index + 1);
    return max(numbers[index], max_val);
}

/**
 * @brief Run the recursive maximum-search demonstration.
 * @return int Exit status code returned to the operating system.
 */
int main() {
    vector<int> numbers = {5, 10, 3, 8, 1};
    cout << find_max(numbers) << '\n';
    return 0;
}
